mrmustard/physics/representations.py

```diff
@@ -0,0 +1,173 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import numpy as np
+
+from abc import ABC, abstractmethod
+from typing import Any, Union
+
+from mrmustard.math import Math
+
+math = Math()
+
+from mrmustard.utils.typing import Scalar, Batch, Matrix, Vector
+from mrmustard.utils import settings
+
+
+class Data(ABC):
+    r"""Abstract parent class for types of data encoding a quantum state's representation."""
+
+    @abstractmethod
+    def __neg__(self) -> Data:
+        ...
+
+    @abstractmethod
+    def __eq__(self, other: Data) -> bool:
+        ...
+
+    @abstractmethod
+    def __add__(self, other: Data) -> Data:
+        ...
+
+    def __sub__(self, other: Data) -> Data:
+        try:
+            return self.__add__(-other)
+        except AttributeError as e:
+            raise TypeError(f"Cannot subtract {self.__class__} and {other.__class__}.") from e
+
+    @abstractmethod
+    def __call__(self, dom: Any) -> Scalar:
+        r"""Evaluate the function at a point in the domain."""
+        ...
+
+    @abstractmethod
+    def __truediv__(self, other: Union[Scalar, Data]) -> Data:
+        ...
+
+    @abstractmethod
+    def __mul__(self, other: Union[Scalar, Data]) -> Data:
+        ...
+
+    def __rmul__(self, other: Scalar) -> Data:
+        return self.__mul__(other=other)
+
+
+class MatVecData(Data):  # Note: this class is abstract!
+    r"""Contains matrix and vector -like data for certain Representation objects.
+
+    Args:
+        mat (Batch[Matrix]): the matrix-like data to be contained in the class
+        vec (Batch[Vector]):    the vector-like data to be contained in the class
+        coeffs (Batch[Scalar]): the coefficients
+    """
+
+    def __init__(self, mat: Batch[Matrix], vec: Batch[Vector], coeffs: Batch[Scalar]) -> None:
+        if coeffs is None:  # default all 1s
+            coeffs = math.ones(len(vec), dtype=math.float64)
+
+        self.mat = math.atleast_3d(math.astensor(mat))
+        self.vec = math.atleast_2d(math.astensor(vec))
+        self.coeffs = math.atleast_1d(math.astensor(coeffs))
+        assert (
+            len(self.mat) == len(self.vec) == len(self.coeffs)
+        ), "All inputs must have the same batch size."
+        assert (
+            self.mat.shape[-1] == self.mat.shape[-2] == self.vec.shape[-1]
+        ), "A and b must have the same dimension and A must be symmetric"
+        self.batch_dim = self.mat.shape[0]
+        self.dim = self.mat.shape[-1]
+
+    def __neg__(self) -> MatVecData:
+        return self.__class__(self.mat, self.vec, -self.coeffs)
+
+    def __eq__(self, other: MatVecData, exclude_scalars: bool = False) -> bool:
```

Comment:
I never defined `__eq__` with more parameters than `other`, because then it's tricky to use `==`. Is `exclude_scalars` absolutely necessary?